Add support for contexts via the unstable GraphQL API

bendrucker opened the pull request:
This adds `circleci_context`, `data.circleci_context`, and `circleci_context_environment_variable` for managing contexts and their environment variables. It integrates the GraphQL client from the CircleCI CLI. While this is admittedly sketchy, the module is vendored and this feature can be labeled experimental. 

Still to do:

* Write and run acceptance tests
* Write some docs (repo-wide, I guess)

Some general notes:

* The APIs for working with contexts are limited (list, create, delete) and very GraphQL-y so we need some extra client logic.
* I tried to avoid adding to what is now `Client`, previously `ProviderClient`. I did define some new functions for working with contexts, but they accept the required client (GraphQL) as an argument and are defined in their own file.
* Imports work with context names and IDs, have their own state functions, and use `/` to delimit identifiers since that won't be a valid character in any of them. Importing an environment variable also requires the value to be provided as an environment variable to Terraform, so it can detect changes. These changes are probably worth porting to the `circleci_environment_variable` resource someday.

Closes #18

bendrucker commented:
Well on the way! Acceptance tests for `circleci_context` are written and passing. Will quickly knock out acceptance tests for the data source as well and then move on to `circleci_context_environment_variable` later today.

```
$ TESTARGS="-run TestAccCircleCIContext" make testacc

TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run TestAccCircleCIContext -timeout 15m
?   	github.com/mrolla/terraform-provider-circleci	[no test files]
=== RUN   TestAccCircleCIContext_basic
--- PASS: TestAccCircleCIContext_basic (5.33s)
=== RUN   TestAccCircleCIContext_update
--- PASS: TestAccCircleCIContext_update (9.12s)
PASS
ok  	github.com/mrolla/terraform-provider-circleci/circleci	14.699s
```

bendrucker commented:
Full acceptance tests going now:

```
$ TESTARGS="-run TestAccCircleCIContext" make testacc
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run TestAccCircleCIContext -timeout 15m
?   	github.com/mrolla/terraform-provider-circleci	[no test files]
=== RUN   TestAccCircleCIContextDataSource
--- PASS: TestAccCircleCIContextDataSource (6.23s)
=== RUN   TestAccCircleCIContextEnvironmentVariable_basic
--- PASS: TestAccCircleCIContextEnvironmentVariable_basic (6.46s)
=== RUN   TestAccCircleCIContextEnvironmentVariable_update
--- PASS: TestAccCircleCIContextEnvironmentVariable_update (9.81s)
=== RUN   TestAccCircleCIContext_basic
--- PASS: TestAccCircleCIContext_basic (4.09s)
=== RUN   TestAccCircleCIContext_update
--- PASS: TestAccCircleCIContext_update (8.26s)
PASS
ok  	github.com/mrolla/terraform-provider-circleci/circleci	35.097s
```

Probably only a few hours work total left to get this out the door. Still to do:

* Tests for imports
* Tests where the org is set as a resource attribute rather than via the provider
* Docs

bendrucker commented:
@mrolla I think this is in a good place for you to take a look. Lots of changes here, take your time!

mrolla commented:
Hey @bendrucker, amazing work! I'm starting to review, but I was wondering if you could split the PR in two and move the website stuff to another PR, to reduce a little bit the scope.

bendrucker commented on GNUmakefile:
```diff
@@ -0,0 +1,62 @@
+errcheck:
+	@sh -c "'$(CURDIR)/scripts/errcheck.sh'"
+
+vendor-status:
```
Sorry yes this is all "standard" stuff that comes that is used as boilerplate for all repos in the `terraform-providers`. I can remove the bits that are definitely not relevant.

bendrucker commented on circleci/contexts.go:
```diff
@@ -0,0 +1,66 @@
+package circleci
```
Yes, meant to provide some notes on that. A centralized client abstraction is not something you'll see in a typical Terraform providers. I understand the temptation given lack of well-supported API clients for Circle, but it's going to break down with more resources. Most of these methods are only used once and are providing little more than a `organization` default, and even that is [partially repeated](https://github.com/mrolla/terraform-provider-circleci/blob/f672452e22c8c4c691e53557fb04f0c70c400bac/circleci/resource_circleci_environment_variable.go#L248-L256).

Individuals functions enforce some behavioral simplicity. They are purely smoothing over a lack of functionality in the API client library and could be removed if a more complete library were published. They can't read from the provider configuration. And the argument type tells you which client (REST vs GraphQL) is used.

To your question about `client_context.go`, yes that is definitely a good move, I'll do that. Seems like all the `client.EnvVar*` methods can just be inlined, happy to do that in a follow up PR.

bendrucker commented:
Thank you for taking a look at this! Will get the conflicts resolved and pull the website stuff out. I need to do some research on how best to deploy that for a community provider.

I've used that setup plenty in official provider repos and didn't realize that every provider is registered as a submodule to make that work:

https://github.com/hashicorp/terraform-website#living-with-submodules

Seems like it should be possible to follow the [new provider repositories](https://github.com/hashicorp/terraform-website#new-provider-repositories) instructions, build to static files, and then publish to GitHub pages.

Leaning towards just doing markdown for now.

bendrucker commented:
Ok, since last time:

* Removed the `website/` content, can tackle general docs separately
* Renamed `contexts.go`
* Removed some irrelevant provider boilerplate
* Merged with master and updated my code to use terraform-plugin-sdk (thanks for that btw!)

Ready for another look!

bendrucker commented:
Awesome, will give this some more thorough testing this week!
